source/src/protocols/drug_design/ReactionBasedAnalogSampler.hh

```diff
@@ -0,0 +1,174 @@
+	/// @brief Draw samples given lists of fragment similarity to first round product reagents
+	utility::vector1< Product > sample_fragment( utility::vector0< std::list< std::pair< core::Real, core::Size > > > & score_idx_set ) const;
+	/// @brief Draw a product from the given candidate set
+	Product sample_candidate( const utility::vector1< Product > & candidate_set, const ::RDKit::RWMOL_SPTR rdmol );
+
+	/// @brief A helper function to easily switch between using ECFP and FCFP
+	::RDKit::SparseIntVect<unsigned int>* getMorganFingerprint( ::RDKit::ROMol const & mol ) const;
+
+	/// @brief Find a common substructure mapping using RDKit Open3DAlign method
+	core::chemical::IndexIndexMapping find_O3A_mapping( ::RDKit::ROMOL_SPTR from, ::RDKit::ROMOL_SPTR to ) const;
+
+	/// @brief Setup weights for a sampler. Weighted sampling with geometric distribution; common ratio dynamically decided by population size
+	void geometric_sampling( numeric::random::WeightedSampler & sampler, core::Size N );
+
+	/// @brief Return analogs of the chosen product, can be itself
+	utility::vector1< Product > analog_search( const ReactionBasedAnalogSampler::Product & prod ) const;
+
+	/// @brief Find all pair of reagents that can undergo the specific reaction from the current sets.
+	void pair( core::Size r_no, utility::vector1< utility::vector0< core::Size > > const & sets, utility::vector1< Product > & candidates, int & revisits ) const;
+	/// @brief Helper function for pairing. Use a DFS to generate all pairs.
+	void pair( core::Size curr_no, core::Size r_no, utility::vector0< core::Size > single_pair, utility::vector0< utility::vector0< core::Size > > & all_pairs, utility::vector1< utility::vector0< core::Size > > const & sets ) const;
+
+	/// @brief Run the reaction given the list of reagent indices
+	Product run_reaction( utility::vector0< core::Size > const & reagents ) const;
+
+	/// @brief helper function for comparing score_index paires
+	static bool sortbySim( const std::pair< core::Real, core::Size >& a, const std::pair< core::Real, core::Size >& b );
+
+private:
+
+	// Products that have been visited in previous MC loops
+	std::unordered_set< Product, Product::HashFunction > visited_;
+
+	// history parameter which records the input structure from last MC cycle
+	std::string last_smiles_;
+
+	// The reactions objects
+	std::unordered_map< std::string, ChemicalReactionOP > rxns_;
+	utility::vector1< Reagent > reagents_;
+	numeric::random::WeightedSampler reagent_sampler;
+
+	// parameter for geometric sampling that mimics the Pareto principle. Ex. ratio = 0.2 means first 20% of the fragments accounts for 80% of the weights.
+	// Should not exceeds 0.25 at any point so that the weight still sums to 1
+	core::Real geo_spl_ratio;
+	// parameters for dynamic sampling, in the order: min, max, step, OFF_after_n_step, base
+	bool dynamic_sampling_;
+	utility::vector0< core::Real > dynamic_spl_ratios;
```

Comment:
Here and elsewhere, for consistency with the rest of Rosetta, prefer using vector1 over vector0 unless there's a specific reason why indexing starting with 0 is required. (And if there is, leave a comment about that reason, such that people in the future know why you're special casing things.)